### ISSUE TITLE
mapper_parsing_exception with elasticsearch-connector

### ISSUE THREAD

**Issue by sr-suyash** [confluentinc/confluent-kafka-python#171]
I am using elasticsearch-kafka-connector with avro schema and pushing the data to a topic. I am trying to push data to elasticsearch in batches, so I have list of dictionaries containing key: value as per defined in schema. Also, I am using confluent-kafka-python as the python client for kafka-connect

The avro schema looks like:
```
{
"name": "dictarray",
"type": "array",
"items": {
    "type": "record",
    "name": "log_fields",
    "fields": [
        {"name": "accept_lang", "type": "string"},
            ....
            ....
            ....

    ]
  }
}
```
For now I am using connect-avro-standalone schema-registry and publishing data to the schema.registry.url as mentioned in confluent-kafka-python user guide and getting the following exception:

[{"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"not_x_content_exception","reason":"Compressor detection can only be called on some xcontent bytes or compressed xcontent bytes"}}]
Need some help here...!!!

**Comment by ewencp** [comment]
@sr-suyash Can you clarify what commands you are actually running? It sounds like the elasticsearch connector is throwing the error. Have you verified that the data is correctly produced from confluent-kafka-python into Kafka? You could use kafka-avro-console-consumer to validate the data in the topic is as expected.

**Comment by sr-suyash** [comment]
@ewencp Thanks for reverting. Here is what I am doing:
I have a requirement where I am reading data stream from a source and I want to do a bulk insert in elasticsearch through connector. As in my previous post, my schema works fine and I verified it from kafka-avro-console-consumer, I am getting an array of dictionaries containing my keys and value as per defined in my avsc file:
```
kafka-avro-console-consumer message:

    [{'accept_lang': '-', ..., 'waf_info': '-', 'xfer_time': '0'},
     {'accept_lang': '-', ..., 'waf_info': '-', 'xfer_time': '0'},
     {'accept_lang': '-', ..., 'waf_info': '-', 'xfer_time': '0'}]
``` 

Here is how my producer config looks like:
```
producerConf = {
    'bootstrap.servers': 'localhost:9092,
    'schema.registry.url': 'http://localhost:8081',
}
```

Here is my code snippet where I have a list in which I am appending streams and sending it as param to avroProducer. 
```
if len(chunk) == 3:
    avroProducer.produce(topic='bulk_test', value=chunk)
    avroProducer.poll(0.05)
```

Another thing I would like to bring, when I am setting poll(0), I am experiencing data drops in elasticsearch when I have a huge stream, say 100,000 and to avoid data drops I am trying to publish data in bulks so that even if I have to poll for for 500ms, it's not impacting the performance (please tell me if I am doing something wrong here as I am fairly new to this)

I was initially suspecting, may be value has to be given as dictionary in avroProducer which is causing the issue, until I saw the correct messages being published from kafka-avro-console-consumer (thank you for pointing that out).

I still got the same error when my connector is trying to sink data in elasticsearch:
```
[2017-04-08 06:20:00,787] WARN Failed to flush during stop (io.confluent.connect.elasticsearch.ElasticsearchWriter:228)
org.apache.kafka.connect.errors.ConnectException: Bulk request failed: [{"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"not_x_content_exception","reason":"Compressor detection can only be called on some xcontent bytes or compressed xcontent bytes"}}]
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.execute(BulkProcessor.java:353)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:326)
	at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:312)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

And here is the mapping getting created for the index:
```
{
   "bulk_test": {
      "mappings": {
         "kafka-connect": {
            "properties": {
               "accept_lang": {
                  "type": "string"
                  },
               ...
               ...
               ...
               }
           }
        }
     }
}
```

**Comment by ewencp** [comment]
@sr-suyash I see. The connector doesn't expect individual Kafka messages to contain multiple documents for elasticsearch. Rather, we expect that each Kafka message corresponds to a single Kafka document and should be compatible with the mapping for the index.

I'm not sure if we can safely allow for bulk updates as I think it might be valid to have an array as the document type. You might want the mapping to specify a [nested type](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html) so each of the elements can be indexed separately.

**Comment by zzbennett** [comment]
I am experiencing a similar issue. I am seeing this upon indexing any record into ES: 
```
[BulkProcessor@2073013577-1] WARN io.confluent.connect.elasticsearch.bulk.BulkProcessor - Failed to execute batch 1 of 1 records, retrying after 100 ms
org.apache.kafka.connect.errors.ConnectException: Bulk request failed: [{"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"not_x_content_exception","reason":"not_x_content_exception: Compressor detection can only be called on some xcontent bytes or compressed xcontent bytes"}}]
        at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.execute(BulkProcessor.java:353)
        at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:326)
        at io.confluent.connect.elasticsearch.bulk.BulkProcessor$BulkTask.call(BulkProcessor.java:312)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

I'm running confluent 3.2.0. I just want the connector to take the incoming payloads (which are JSON) and index them as is into ES, so I have set `schema.ignore.config=true`. I have also specified `key.converter.schemas.enable=false` and `value.converter.schemas.enable=false`.  I've added some debug info to see what is going wrong and I noticed the payload for the IndexableRecord is an escaped json String, i.e.
```
"{\"entry_point\": \"rabbitmq\", \"ts_posted\": \"2017-04-05T02:02:27.296757\", \"tracer_id\": \"eb1c8284-19a3-11e7-9555-0a7b84c6deb8\"}
```

The value in the SinkRecord is a non-escaped String containing the json data.

ES is expecting this to be a JSON payload (without the escaping). 
While debugging, I wrote a unit test to reproduce this. In ElasticsearchWriterTest, the test looks like this:

```
  @Test
  public void testWriterIgnoreSchemaStringRecord() throws Exception {
    final boolean ignoreKey = true;
    final boolean ignoreSchema = true;

    String value = "{\"foo\":\"bar\"}";
    Collection<SinkRecord> records = Collections.singletonList(new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, key, Schema.STRING_SCHEMA, value, 0));
    ElasticsearchWriter writer = initWriter(client, ignoreKey, ignoreSchema);
    writeDataAndRefresh(writer, records);
    verifySearchResults(records, ignoreKey, ignoreSchema);
  }
```

I haven't dug further yet to figure out why this is happening, but I figured I'd post what I have found so far.

**Comment by ewencp** [comment]
@zzbennett Cool, thanks for digging into that a bit. The escaped JSON sounds a lot like the wrong Converter is being used -- this is the behavior I would expect to see if the data was being read from Kafka using StringConverter instead of JsonConverter. What do your Connect worker configs look like? And is the converter being overridden for the connector? And I'm guessing you've already verified that the data in Kafka looks like you would expect it to, e.g. by using the console consumer to inspect a few records?

**Comment by zzbennett** [comment]
Hey @ewencp. I do think the wrong converter is being used. There's code in the elasticsearch connector that uses the JsonConverter when schemas are disabled which I think is the problem. I actually didn't notice this issue had been filed under the confluent-kafka-python project when I commented. It's an issue with the elasticsearch connector. I've created a separate issue in that project here: https://github.com/confluentinc/kafka-connect-elasticsearch/issues/82
